### PR TITLE
Bug/lwip tunnel testing

### PR DESCRIPTION
**Pull request by robszewczyk**
Initial groundwork for enabling tunnel testing under LwIP: this pull request cleans up the relations between network and weave configurations, adds the ability to have LwIP listen on multiple interfaces, and to pick up configuration of those interfaces from the underlying tap devices.   This does not fully address the needs of the LwIP-based tunnel testing; the following items remain unaddressed:

* mock-border-gw does not advertise routes
* test scripts remain broken, in particular w.r.t. the conditions they use to detect progress of program under test.  

Nonetheless, given that each of the commits is getting somewhat large, the commits remain self-contained and could be merged for an incremental win

### CODE REVIEW

File: src/lwip/standalone/TapInterface.c [hunk 1]
```diff
@@ -458,6 +458,26 @@ TapInterface_Init(TapInterface *tapif, const char *interfaceName, u8_t *macAddr)
         perror("TapInterface: ioctl(TUNSETIFF) failed");
         return ERR_IF;
     }
+
+#if 0
```
Comment (gerickson) [inline src/lwip/standalone/TapInterface.c]: Please feature-flag this according rather than #if 0ing it.

Comment (robszewczyk) [reply src/lwip/standalone/TapInterface.c]: I will remove this.  I thought this was a handy note about how to fully mirror interface configuration into LwIP, but at this time it feels like the TAP interface (or perhaps the ethernet bridge) is far more touchy about the MAC addresses than I expected.  That's why I kept the potentially dodgy address assignment based on the PID -- bizarrely enough, it worked better than a fully random 48 bit address.

File: src/test-apps/TapAddrAutoconf.cpp [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+            if (strcmp(ifName, curAddr->ifa_name) == 0)
+            {
+                char buf[128];
+                const char * rv;
+                char *tmp;
+                rv = inet_ntop(curAddr->ifa_addr->sa_family,
+                               curAddr->ifa_addr->sa_family == AF_INET6 ?
+                               (const void *)&(((sockaddr_in6 *)curAddr->ifa_addr)->sin6_addr) :
+                               (const void *)&(((sockaddr_in *)curAddr->ifa_addr)->sin_addr),
+                               buf, 128);
```
Comment (gerickson) [inline src/test-apps/TapAddrAutoconf.cpp]: 128 -> sizeof (buf)

File: src/test-apps/TapAddrAutoconf.cpp [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    const int err = getifaddrs(&addrsList);
+
+    if (err == 0)
+    {
+        curAddr = addrsList;
+        while (curAddr)
+        {
+            if (strcmp(ifName, curAddr->ifa_name) == 0)
+            {
+                char buf[128];
```
Comment (gerickson) [inline src/test-apps/TapAddrAutoconf.cpp]: 128 -> Magic Number? This should probably be INET6_ADDRSTRLEN.

File: src/test-apps/TestInetLayer.cpp [hunk 4]
```diff
@@ -935,16 +935,20 @@ static void StartTest(void)
     IPAddressType      lIPAddressType = kIPAddressType_IPv6;
     IPProtocol         lIPProtocol    = kIPProtocol_ICMPv6;
     IPVersion          lIPVersion     = kIPVersion_6;
-    IPAddress          lAddress       = gNetworkOptions.LocalIPv6Addr;
+    IPAddress          lAddress       = nl::Inet::IPAddress::Any;
     INET_ERROR         lStatus;
```
Comment (gerickson) [inline src/test-apps/TestInetLayer.cpp]: Style: Newline after this line.

File: src/test-apps/TestInetLayer.cpp [hunk 5]
```diff
@@ -935,16 +935,20 @@ static void StartTest(void)
...
-
+    if (!gNetworkOptions.LocalIPv6Addr.empty())
+        lAddress = gNetworkOptions.LocalIPv6Addr[0];
 #if INET_CONFIG_ENABLE_IPV4
     if (gOptFlags & kOptFlagUseIPv4)
     {
         lIPAddressType = kIPAddressType_IPv4;
         lIPProtocol    = kIPProtocol_ICMPv4;
         lIPVersion     = kIPVersion_4;
```
Comment (gerickson) [inline src/test-apps/TestInetLayer.cpp]: Style: new line after this line.

File: src/test-apps/ToolCommon.cpp [hunk 6]
```diff
@@ -39,7 +39,7 @@
 #include <sys/types.h>
 #include <unistd.h>
 #include <new>
-
+#include <vector>
```
Comment (gerickson) [inline src/test-apps/ToolCommon.cpp]: Is ToolCommon ever included in on-device (RTOS) tests?

File: src/test-apps/ToolCommon.cpp [hunk 7]
```diff
@@ -874,10 +952,15 @@ void InitNetwork()
...
     // Force new IP address to be ready, bypassing duplicate detection.
 
-    if (gNetworkOptions.LocalIPv6Addr != IPAddress::Any) {
-      netif_ip6_addr_set_state(&netIF, 2, 0x30);
-    } else {
-      netif_ip6_addr_set_state(&netIF, 1, 0x30);
+    for (size_t j = 0; j < gNetworkOptions.TapDeviceName.size(); j++)
+    {
+        if (j < gNetworkOptions.LocalIPv6Addr.size()) {
+            netif_ip6_addr_set_state(&(netIFs[j]), 2, 0x30);
```
Comment (gerickson) [inline src/test-apps/ToolCommon.cpp]: Magic number?

File: src/test-apps/ToolCommon.cpp [hunk 8]
```diff
@@ -874,10 +952,15 @@ void InitNetwork()
...
+        }
+        else
+        {
+            netif_ip6_addr_set_state(&(netIFs[j]), 1, 0x30);
+        }
     }
 
 #if INET_CONFIG_ENABLE_DNS_RESOLVER
```
Comment (gerickson) [inline src/test-apps/ToolCommon.cpp]: Magic number?

File: src/test-apps/ToolCommonOptions.h [hunk 9]
```diff
@@ -25,6 +25,8 @@
 #ifndef TOOLCOMMONOPTIONS_H_
 #define TOOLCOMMONOPTIONS_H_
 
+#include <vector>
```
Comment (gerickson) [inline src/test-apps/ToolCommonOptions.h]: Is this code ever included in RTOS-based, on-device tests?

**Comment by gerickson** [comment]
Overall, seems reasonable. I left a few minor comments to address. In the macro perspective, it seems like this all could use a high-level comment somewhere (not exactly clear where) as to the approach and strategy adopted.

File: src/test-apps/ToolCommon.cpp (hunk 6 above)
Comment (robszewczyk) [reply src/test-apps/ToolCommon.cpp]: No, its not.

File: src/lwip/standalone/TapInterface.c (hunk 1 above)
Comment (robszewczyk) [reply src/lwip/standalone/TapInterface.c]: done

File: src/test-apps/TapAddrAutoconf.cpp (hunk 2 above)
Comment (robszewczyk) [reply src/test-apps/TapAddrAutoconf.cpp]: done

File: src/test-apps/TapAddrAutoconf.cpp (hunk 3 above)
Comment (robszewczyk) [reply src/test-apps/TapAddrAutoconf.cpp]: done

File: src/test-apps/TestInetLayer.cpp (hunk 4 above)
Comment (robszewczyk) [reply src/test-apps/TestInetLayer.cpp]: done

File: src/test-apps/TestInetLayer.cpp (hunk 5 above)
Comment (robszewczyk) [reply src/test-apps/TestInetLayer.cpp]: done

File: src/test-apps/ToolCommonOptions.h (hunk 9 above)
Comment (robszewczyk) [reply src/test-apps/ToolCommonOptions.h]: It is not.

**Pull request merged**